### ISSUE TITLE
Model does not learn with u-mup

### ISSUE THREAD

**Issue by SwayStar123** [graphcore-research/unit-scaling#88]
Im training a diffusion transformer, decided to try to implement u-mup into it however spending hours scaling all the forwards and backwards to try to get it to ~std 1 for all activations, gradients, and weights, it just doesnt seem to be learning!

Activations:
```
siglip_norm.act.std = 1.00                  
siglip_embedder.mlp.linear.act.std = 1.00        
time_embedder.mlp.mlp.0.linear.act.std = 1.49    -> Fixed with scale_fwd
time_embedder.mlp.mlp.2.linear.act.std = 1.43    -> Fixed with scale_fwd
vec_norm.act.std = 1.00                  
vector_embedder.mlp.linear.act.std = 0.99        
image_embedder.linear.act.std = 1.00             
backbone.layers.0.norm1.act.std = 1.00           
backbone.attn.qkv.act.std = 1.00                 
backbone.attn.proj.act.std = 3.67                 -> Fixed with scale_fwd
backbone.layers.0.norm2.act.std = 1.00           
backbone.layers.0.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.0.mlp.mlp.2.linear.act.std = 1.01
backbone.layers.1.norm1.act.std = 1.00           
backbone.layers.1.norm2.act.std = 1.00           
backbone.layers.1.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.1.mlp.mlp.2.linear.act.std = 1.01
backbone.layers.2.norm1.act.std = 1.00           
backbone.layers.2.norm2.act.std = 1.00           
backbone.layers.2.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.2.mlp.mlp.2.linear.act.std = 1.00
backbone.layers.3.norm1.act.std = 1.00           
backbone.layers.3.norm2.act.std = 1.00           
backbone.layers.3.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.3.mlp.mlp.2.linear.act.std = 1.00
backbone.layers.4.norm1.act.std = 1.00           
backbone.layers.4.norm2.act.std = 1.00           
backbone.layers.4.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.4.mlp.mlp.2.linear.act.std = 1.00
backbone.layers.5.norm1.act.std = 1.00           
backbone.layers.5.norm2.act.std = 1.00           
backbone.layers.5.mlp.mlp.0.linear.act.std = 1.00
backbone.layers.5.mlp.mlp.2.linear.act.std = 1.01
output_layer.norm_final.act.std = 1.00      
output_layer.mlp.mlp.0.linear.act.std = 1.00
output_layer.mlp.mlp.2.linear.act.std = 1.00
```

Weights are all 1s and 0s ofcourse

```
time_embedder.mlp.mlp.0.linear.weight.std = 1.00                                                      
time_embedder.mlp.mlp.0.linear.bias.std = 0.00                                                        
time_embedder.mlp.mlp.2.linear.weight.std = 1.00 
time_embedder.mlp.mlp.2.linear.bias.std = 0.00   
image_embedder.linear.weight.std = 1.00          
siglip_embedder.mlp.linear.weight.std = 1.00     
siglip_embedder.mlp.linear.bias.std = 0.00       
vector_embedder.mlp.linear.weight.std = 1.00     
vector_embedder.mlp.linear.bias.std = 0.00       
backbone.attn.qkv.weight.std = 1.00                                                                   
backbone.attn.proj.weight.std = 1.00                                                                  
backbone.layers.0.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.0.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.0.mlp.mlp.2.linear.weight.std = 1.00     
backbone.layers.0.mlp.mlp.2.linear.bias.std = 0.00                                                    
backbone.layers.1.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.1.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.1.mlp.mlp.2.linear.weight.std = 1.00     
backbone.layers.1.mlp.mlp.2.linear.bias.std = 0.00                                                    
backbone.layers.2.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.2.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.2.mlp.mlp.2.linear.weight.std = 1.00     
backbone.layers.2.mlp.mlp.2.linear.bias.std = 0.00                                                    
backbone.layers.3.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.3.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.3.mlp.mlp.2.linear.weight.std = 1.00     
backbone.layers.3.mlp.mlp.2.linear.bias.std = 0.00                                                    
backbone.layers.4.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.4.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.4.mlp.mlp.2.linear.weight.std = 1.00     
backbone.layers.4.mlp.mlp.2.linear.bias.std = 0.00                                                    
backbone.layers.5.mlp.mlp.0.linear.weight.std = 1.00     
backbone.layers.5.mlp.mlp.0.linear.bias.std = 0.00                                                    
backbone.layers.5.mlp.mlp.2.linear.weight.std = 1.00
backbone.layers.5.mlp.mlp.2.linear.bias.std = 0.00
output_layer.mlp.mlp.0.linear.weight.std = 1.00
output_layer.mlp.mlp.0.linear.bias.std = 0.00
output_layer.mlp.mlp.2.linear.weight.std = 1.01
output_layer.mlp.mlp.2.linear.bias.std = 0.00
```

The gradients are a bit more skewed around but I tried to make sure theyre all near 1 as best i could
```
image_embedder.linear.weight.grad.std = 1.62
siglip_embedder.mlp.linear.weight.grad.std = 1.49
siglip_embedder.mlp.linear.bias.grad.std = 5.48
vector_embedder.mlp.linear.weight.grad.std = 1.51
vector_embedder.mlp.linear.bias.grad.std = 1.39
backbone.attn.qkv.weight.grad.std = 3.15
backbone.attn.proj.weight.grad.std = 5.43
backbone.layers.0.mlp.mlp.0.linear.weight.grad.std = 1.33
backbone.layers.0.mlp.mlp.0.linear.bias.grad.std = 10.02
backbone.layers.0.mlp.mlp.2.linear.weight.grad.std = 1.32
backbone.layers.0.mlp.mlp.2.linear.bias.grad.std = 10.09
backbone.layers.1.mlp.mlp.0.linear.weight.grad.std = 1.00
backbone.layers.1.mlp.mlp.0.linear.bias.grad.std = 4.99
backbone.layers.1.mlp.mlp.2.linear.weight.grad.std = 1.00
backbone.layers.1.mlp.mlp.2.linear.bias.grad.std = 4.95
backbone.layers.2.mlp.mlp.0.linear.weight.grad.std = 0.81
backbone.layers.2.mlp.mlp.0.linear.bias.grad.std = 2.56
backbone.layers.2.mlp.mlp.2.linear.weight.grad.std = 0.81
backbone.layers.2.mlp.mlp.2.linear.bias.grad.std = 2.54
backbone.layers.3.mlp.mlp.0.linear.weight.grad.std = 0.73
backbone.layers.3.mlp.mlp.0.linear.bias.grad.std = 1.39
backbone.layers.3.mlp.mlp.2.linear.weight.grad.std = 0.73
backbone.layers.3.mlp.mlp.2.linear.bias.grad.std = 1.37
backbone.layers.4.mlp.mlp.0.linear.weight.grad.std = 0.68
backbone.layers.4.mlp.mlp.0.linear.bias.grad.std = 0.82
backbone.layers.4.mlp.mlp.2.linear.weight.grad.std = 0.69
backbone.layers.4.mlp.mlp.2.linear.bias.grad.std = 0.80
backbone.layers.5.mlp.mlp.0.linear.weight.grad.std = 0.68
backbone.layers.5.mlp.mlp.0.linear.bias.grad.std = 0.60
backbone.layers.5.mlp.mlp.2.linear.weight.grad.std = 0.68
backbone.layers.5.mlp.mlp.2.linear.bias.grad.std = 0.60
output_layer.mlp.mlp.0.linear.weight.grad.std = 0.70
output_layer.mlp.mlp.0.linear.bias.grad.std = 0.63
output_layer.mlp.mlp.2.linear.weight.grad.std = 0.70
output_layer.mlp.mlp.2.linear.bias.grad.std = 0.63
```

I am using the uu optim AdamW, and I am using the libraries mse loss function, so Im not sure what else there is to fix now.

For comparison, Image generated with model trained using u-mup:

![Image](https://github.com/user-attachments/assets/1f3ded43-83d4-455e-9b9d-d45f9d129149)
 
And image generated using model trained with SP:

![Image](https://github.com/user-attachments/assets/972a6fe1-7a41-414c-8d6b-35178db5ae71)

For same number of train steps. 

I have tried changing the LR between values of 0.0001 to 1, none of it helps. I am at a loss as to what else to change, any help would be appreciated.
@DouglasOrr

**Comment by DouglasOrr** [comment]
Thanks for getting in touch - sorry that it didn't just work.

My initial thought was regarding learning rate, since that generally needs increasing for u-muP, as you've already tested. What do the training loss curves look like as you increase LR, especially compared against the working baseline?

One thing that might be useful is to try to track down the actual (after scaling) learning rate applied to each trainable parameter, to make sure that none of them are too extreme.

One place we've seen adverse effects of learning rate differences is in vector parameters e.g. biases & norm affine parameters - these can get a higher effective LR under u-muP than SP, so if non-trainable biases works on the baseline that might be something to test.
